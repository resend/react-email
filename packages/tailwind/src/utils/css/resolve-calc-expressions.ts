import { type CssNode, walk } from 'css-tree';

/**
 * Intentionally only resolves `*` and `/` operations without dealing with parenthesis, because this is the only thing required to run Tailwind v4
 */
export function resolveCalcExpressions(node: CssNode) {
  walk(node, {
    visit: 'Function',
    enter(func, funcListItem) {
      if (func.name === 'calc') {
        /*
          [
            { type: 'Dimension', loc: null, value: '0.25', unit: 'rem' },
            { type: 'Operator', loc: null, value: '*' },
            { type: 'Number', loc: null, value: '2' }
            { type: 'Percentage', loc: null, value: '2' }
          ]
        */
        func.children.forEach((child, item) => {
          const left = item.prev;
          const right = item.next;
          if (
            left &&
            right &&
            child.type === 'Operator' &&
            (left.data.type === 'Dimension' ||
              left.data.type === 'Number' ||
              left.data.type === 'Percentage') &&
            (right.data.type === 'Dimension' ||
              right.data.type === 'Number' ||
              right.data.type === 'Percentage')
          ) {
            if (child.value === '*' || child.value === '/') {
              const value = (() => {
                if (child.value === '*') {
                  return String(
                    Number.parseFloat(left.data.value) *
                    Number.parseFloat(right.data.value),
                  );
                }
                if (right.data.value === '0') {
                  return '0';
                }
                return String(
                  Number.parseFloat(left.data.value) /
                  Number.parseFloat(right.data.value),
                );
              })();
              if (
                left.data.type === 'Dimension' &&
                right.data.type === 'Number'
              ) {
                item.data = {
                  type: 'Dimension',
                  unit: left.data.unit,
                  value,
                };
                func.children.remove(left);
                func.children.remove(right);
              } else if (
                left.data.type === 'Number' &&
                right.data.type === 'Dimension'
              ) {
                item.data = {
                  type: 'Dimension',
                  unit: right.data.unit,
                  value,
                };
                func.children.remove(left);
                func.children.remove(right);
              } else if (
                left.data.type === 'Number' &&
                right.data.type === 'Number'
              ) {
                item.data = {
                  type: 'Number',
                  value,
                };
                func.children.remove(left);
                func.children.remove(right);
              } else if (
                left.data.type === 'Dimension' &&
                right.data.type === 'Dimension' &&
                left.data.unit === right.data.unit
              ) {
                if (child.value === '/') {
                  item.data = {
                    type: 'Number',
                    value,
                  };
                } else {
                  item.data = {
                    type: 'Dimension',
                    unit: left.data.unit,
                    value,
                  };
                }
                func.children.remove(left);
                func.children.remove(right);
              } else if (
                left.data.type === 'Percentage' &&
                right.data.type === 'Number'
              ) {
                item.data = {
                  type: 'Percentage',
                  value,
                };
                func.children.remove(left);
                func.children.remove(right);
              } else if (
                left.data.type === 'Number' &&
                right.data.type === 'Percentage'
              ) {
                item.data = {
                  type: 'Percentage',
                  value,
                };
                func.children.remove(left);
                func.children.remove(right);
              } else if (
                left.data.type === 'Percentage' &&
                right.data.type === 'Percentage'
              ) {
                if (child.value === '/') {
                  item.data = {
                    type: 'Number',
                    value,
                  };
                } else {
                  item.data = {
                    type: 'Percentage',
                    value,
                  };
                }
                func.children.remove(left);
                func.children.remove(right);
              }
            }
          }
        });
        if (func.children.size === 1 && func.children.first) {
          funcListItem.data = func.children.first;
        }
      }
    },
  });
}
