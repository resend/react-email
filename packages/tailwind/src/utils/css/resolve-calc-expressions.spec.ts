import { generate, parse } from 'css-tree';
import { resolveCalcExpressions } from './resolve-calc-expressions';

describe('resolveCalcExpressions()', () => {
  it('resolves spacing calc expressions from tailwind v4', () => {
    const root = parse(`
.px-3{padding-inline:calc(0.25rem*3)}
.py-2{padding-block:calc(0.25rem*2)}
  `);
    resolveCalcExpressions(root);
    expect(generate(root)).toMatchSnapshot();
  });

  it('resolves calc expressions repeating decimals', () => {
    const root = parse(`
      .w-1\/3 { width: calc(0.3333333333333333*100%); }
    `);
    resolveCalcExpressions(root);
    expect(generate(root)).toMatchSnapshot();
  });

  it('does not modify complex calc expressions', () => {
    const root = parse(`
.px-3{padding-inline:calc(0.25rem*(3 + 1px))}
.py-2{padding-block:calc(0.25rem*(2 + 1px))}
  `);
    resolveCalcExpressions(root);
    expect(generate(root)).toMatchSnapshot();
  });
});
