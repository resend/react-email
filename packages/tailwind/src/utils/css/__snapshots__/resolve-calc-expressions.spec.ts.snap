// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolveCalcExpressions() > should not do anything to complex calc expressions 1`] = `".px-3{padding-inline:calc(0.25rem*(3 + 1px))}.py-2{padding-block:calc(0.25rem*(2 + 1px))}"`;

exports[`resolveCalcExpressions() > should resolve spacing calc expressions from tailwind v4 1`] = `".px-3{padding-inline:0.75rem}.py-2{padding-block:0.5rem}"`;
